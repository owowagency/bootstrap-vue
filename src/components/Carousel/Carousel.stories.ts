import Carousel from '.';
import {CarouselSlide} from '@/components/Carousel/Carousel.vue';

const slides: CarouselSlide[] = [
    {
        description: 'Frank the kitten',
        image: 'https://placekitten.com/900/500',
        title: 'Kitten #1',
    },
    {
        description: 'Snowball the kitten',
        image: 'https://placekitten.com/800/450',
        title: 'Kitten #2',
    },
    {
        description: 'Munchie the kitten',
        image: 'https://placekitten.com/850/475',
        title: 'Kitten #3',
    },
];

export default {
    title: 'Components/Carousel',
    component: Carousel,
    argTypes: {
        dark: {
            control: {type: 'boolean'},
        },
        fade: {
            control: {type: 'boolean'},
        },
        hideControls: {
            control: {type: 'boolean'},
        },
        hideIndicators: {
            control: {type: 'boolean'},
        },
        interval: {
            control: {type: 'number'},
        },
    },
};

const template = (args) => ({
    components: {Carousel},
    setup: () => ({args}),
    template: `
        <div class="container">
            <Carousel
                id="Carousel"
                v-bind="args"
            />
        </div>
    `,
});

export const Default = template.bind({});

Default.args = {
    slides,
};
