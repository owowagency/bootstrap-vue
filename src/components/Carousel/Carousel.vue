<template>
    <div
        :id="id"
        ref="carousel"
        class="carousel slide"
        :class="classes"
        :data-bs-interval="interval"
        data-bs-ride="carousel"
    >
        <div class="carousel-inner">
            <slot
                name="slides"
                :slides="slides"
            >
                <div
                    v-for="(slide, index) in slides"
                    :key="`slide-${index}`"
                    class="carousel-item"
                    :class="{active: index === activeIndex}"
                    :data-bs-interval="slide.interval"
                >
                    <slot
                        :index="index"
                        :slide="slide"
                    >
                        <img
                            v-if="slide.image"
                            :src="slide.image"
                            class="d-block w-100"
                        >

                        <slot
                            name="caption"
                            :description="slide.description"
                            :title="slide.title"
                        >
                            <div
                                v-if="slide.title || slide.description"
                                class="carousel-caption d-none d-md-block"
                            >
                                <h5 v-if="slide.title">
                                    {{ slide.title }}
                                </h5>

                                <p v-if="slide.description">
                                    {{ slide.description }}
                                </p>
                            </div>
                        </slot>
                    </slot>
                </div>
            </slot>
        </div>

        <div v-if="!hideControls">
            <button
                class="carousel-control-prev"
                :data-bs-target="`#${id}`"
                data-bs-slide="prev"
            >
                <span class="carousel-control-prev-icon" />
            </button>

            <button
                class="carousel-control-next"
                :data-bs-target="`#${id}`"
                data-bs-slide="next"
            >
                <span class="carousel-control-next-icon" />
            </button>
        </div>

        <div
            v-if="!hideIndicators"
            class="carousel-indicators"
        >
            <button
                v-for="(_, index) in slides"
                :key="`slide-indicator-${index}`"
                :class="{active: index === activeIndex}"
                :data-bs-target="`#${id}`"
                :data-bs-slide-to="index"
            />
        </div>
    </div>
</template>

<script lang="ts">
import {PropType, computed, ref} from 'vue';
import {idProps} from '@/composables/useId';
import useBootstrapEmits from '@/composables/useBootstrapEmits';
import useBootstrapInstance from '@/composables/useBootstrapInstance';
import useClasses from '@/composables/useClasses';

const carouselEvents = ['slid', 'slide'] as const;

type CarouselEvent = typeof carouselEvents[number];

export interface CarouselSlide {
    active?: boolean;
    description?: string;
    image?: string;
    interval?: number;
    title?: string;
}
</script>

<script lang="ts" setup>
const props = defineProps({
    dark: {
        type: Boolean,
        default: false,
    },
    fade: {
        type: Boolean,
        default: false,
    },
    hideControls: {
        type: Boolean,
        default: false,
    },
    hideIndicators: {
        type: Boolean,
        default: false,
    },
    ...idProps,
    interval: {
        type: Number,
        default: 5000,
    },
    slides: {
        type: Array as PropType<CarouselSlide[]>,
        default: () => [],
    },
});

const emit = defineEmits<{(event: CarouselEvent): void}>();

const carousel = ref<HTMLElement>();

const activeIndex = computed(() => {
    const activeSlideIndex = props.slides.findIndex((slide: CarouselSlide) => slide.active);

    return activeSlideIndex > -1 ? activeSlideIndex : 0;
});

useBootstrapEmits(
    carousel,
    carouselEvents,
    emit,
    'carousel',
);

const {bsInstance: bsCarousel} = useBootstrapInstance(
    'Carousel',
    carousel,
);

const {classes} = useClasses(computed(() => [
    props.dark ? 'carousel-dark' : null,
    props.fade ? 'carousel-fade' : null,
]));

defineExpose({bsCarousel});
</script>
