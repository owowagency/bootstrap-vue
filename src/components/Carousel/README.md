# Carousel

```vue
<Carousel :slides="[CarouselSlide]" interval="10000" hideControls />
```

## Imported Props

| Prop name | Description                   | Type   | Values | Default |
| --------- | ----------------------------- | ------ | ------ | ------- |
| id        | The id used for the component | string | -      | `uuid`  |

<!-- TODO Auto-generate based on type? -->

## Events

| Name  | Args | Description                                                 |
| ----- | ---- | ----------------------------------------------------------- |
| slid  |      | Fired when the carousel has completed its slide transition  |
| slide |      | Fires immediately when the slide instance method is invoked |

## Props

| Prop name      | Description                                  | Type            | Values | Default     |
| -------------- | -------------------------------------------- | --------------- | ------ | ----------- |
| dark           | Use dark mode for the controls               | boolean         | -      | false       |
| fade           | Use a fade transition for the slides         | boolean         | -      | false       |
| hideControls   | Hide the controls                            | boolean         | -      | false       |
| hideIndicators | Hide the indicators                          | boolean         | -      | false       |
| interval       | The interval of the carousel in milliseconds | number          | -      | 5000        |
| slides         | The slides of the carousel                   | CarouselSlide[] | -      | () =&gt; [] |

## Slots

| Name         | Description                          | Bindings                                                                                                |
| ------------ | ------------------------------------ | ------------------------------------------------------------------------------------------------------- |
| slides       | Displays content inside the carousel | **slides** `array` - The `CarouselSlide` list                                                           |
| default      | Displays content inside the slide    | **index** `int` - The index of the slide<br/>**slide** `object` - The `CarouselSlide` object            |
| caption      | Display the caption of the carousel  | **description** `string` - The description of the slide<br/>**title** `string` - The title of the slide |
| control-prev | Renders a previous control button    | **dataset** `Object` - Bootstrap data attributes required for controlling the carousel                  |
| control-next | Renders a next control button        | **dataset** `Object` - Bootstrap data attributes required for controlling the carousel                  |
| indicator    | Renders a next control button        | <br/><br/><br/>**dataset** `Object` - Bootstrap data attributes required for controlling the carousel   |
