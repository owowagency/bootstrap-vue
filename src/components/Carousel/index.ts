import Carousel from './Carousel.vue';

export default Carousel;
