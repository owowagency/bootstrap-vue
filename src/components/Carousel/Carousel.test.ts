import {describe, expect, it, vi} from 'vitest';
import Carousel from '.';
import {shallowMount} from '@vue/test-utils';

vi.mock('bootstrap', () => ({
    Carousel: {
        getOrCreateInstance: vi.fn(),
    },
    __v_isShallow: true,
}));

const props = {
    slides: [
        {
            image: 'https://placekitten.com/200/200',
            title: 'Kitten #1',
        },
        {
            active: true,
            image: 'https://placekitten.com/200/200',
            title: 'Kitten #2',
        },
        {
            image: 'https://placekitten.com/200/200',
            interval: 1000,
            title: 'Kitten #3',
        },
    ],
};

describe('template', () => {
    componentRenderTest(Carousel, {
        // Force an id to prevent snapshots from failing because of uuid.
        props: {
            id: 'dont-change-please',
            ...props,
        },
    });

    componentSlotRenderTest(Carousel, 'default', {props});

    componentSlotRenderTest(Carousel, 'slides', {props});

    componentSlotRenderTest(Carousel, 'caption', {props});

    componentSlotRenderTest(Carousel, 'control-prev', {props});

    componentSlotRenderTest(Carousel, 'control-next', {props});

    componentSlotRenderTest(Carousel, 'indicator', {props});

    componentWrapperAttributeTest(Carousel, {props}, 'data-bs-interval', '5000');

    componentWrapperClassTest(Carousel, {dark: true}, 'carousel-dark');

    componentWrapperClassTest(Carousel, {fade: true}, 'carousel-fade');

    ['slid', 'slide'].forEach((event: string) => {
        componentBootstrapEventTest(
            Carousel,
            '.carousel',
            event,
            'carousel',
        );
    });

    it('sets active index from slides', async() => {
        const wrapper = await shallowMount(
            Carousel,
            {props},
        );

        const slides = wrapper.findAll('.carousel-item');

        expect(slides[0].classes()).not.toContain('active');
        expect(slides[1].classes()).toContain('active');
        expect(slides[2].classes()).not.toContain('active');
    });

    it('sets slide interval from slides', async() => {
        const wrapper = await shallowMount(
            Carousel,
            {props},
        );

        const slides = wrapper.findAll('.carousel-item');

        expect(slides[2].attributes('data-bs-interval')).toContain('1000');
    });
});

describe('onMounted', () => {
    it('sets bsCarousel', async() => {
        const wrapper = await shallowMount(Carousel);

        expect((await import('bootstrap')).Carousel.getOrCreateInstance).toBeCalledWith(wrapper.vm.$refs.carousel, {});

        // TODO: Unable to assert bsCarousel value.
    });
});

describe('hideControls', () => {
    it('hides the controls', async() => {
        const wrapper = await shallowMount(
            Carousel,
            {
                props: {
                    hideControls: true,
                    ...props,
                },
            },
        );

        expect(wrapper.find('[data-bs-slide]').exists()).toBeFalsy();
    });
});

describe('hideIndicators', () => {
    it('hides the indicators', async() => {
        const wrapper = await shallowMount(
            Carousel,
            {
                props: {
                    hideIndicators: true,
                    ...props,
                },
            },
        );

        expect(wrapper.find('[data-bs-slide-to]').exists()).toBeFalsy();
    });
});
