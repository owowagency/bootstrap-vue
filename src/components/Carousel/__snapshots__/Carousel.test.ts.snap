// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template renders default 1`] = `
"<div id=\\"dont-change-please\\" class=\\"carousel slide\\" data-bs-interval=\\"5000\\" data-bs-ride=\\"carousel\\">
  <div class=\\"carousel-inner\\">
    <div class=\\"carousel-item\\"><img src=\\"https://placekitten.com/200/200\\" class=\\"d-block w-100\\">
      <div class=\\"carousel-caption d-none d-md-block\\">
        <h5>Kitten #1</h5>
        <!--v-if-->
      </div>
    </div>
    <div class=\\"carousel-item active\\"><img src=\\"https://placekitten.com/200/200\\" class=\\"d-block w-100\\">
      <div class=\\"carousel-caption d-none d-md-block\\">
        <h5>Kitten #2</h5>
        <!--v-if-->
      </div>
    </div>
    <div class=\\"carousel-item\\" data-bs-interval=\\"1000\\"><img src=\\"https://placekitten.com/200/200\\" class=\\"d-block w-100\\">
      <div class=\\"carousel-caption d-none d-md-block\\">
        <h5>Kitten #3</h5>
        <!--v-if-->
      </div>
    </div>
  </div>
  <div><button class=\\"carousel-control-prev\\" data-bs-target=\\"#dont-change-please\\" data-bs-slide=\\"prev\\"><span class=\\"carousel-control-prev-icon\\"></span></button><button class=\\"carousel-control-next\\" data-bs-target=\\"#dont-change-please\\" data-bs-slide=\\"next\\"><span class=\\"carousel-control-next-icon\\"></span></button></div>
  <div class=\\"carousel-indicators\\"><button class=\\"\\" data-bs-target=\\"#dont-change-please\\" data-bs-slide-to=\\"0\\"></button><button class=\\"active\\" data-bs-target=\\"#dont-change-please\\" data-bs-slide-to=\\"1\\"></button><button class=\\"\\" data-bs-target=\\"#dont-change-please\\" data-bs-slide-to=\\"2\\"></button></div>
</div>"
`;
